### ISSUE TITLE
Dataloader not dispatching properly for properties defined using a suspend function

### ISSUE THREAD

**Issue by aravind88** [ExpediaGroup/graphql-kotlin#986]
**Library Version**
`4.0.0-alpha.9` and `3.7.0`

**Describe the bug**
I am fairly new to graphql and `graphql-kotlin`. I have been trying out the library in a ktor server and I noticed that the request was consistently getting stuck (i.e. no response to the client). I was able to debug the issue further and noticed that this only happens when a client requests a field that was provided by a `FunctionDataFetcher`. I tried to verify that it is not an issue on my setup/usage by using the `ktor-server` example project. I noticed the same issue on the example project as well. 

```
// In terminal
git clone git@github.com:ExpediaGroup/graphql-kotlin.git
cd graphql-kotlin/examples
./gradlew server:ktor-server:run
```

```
// In browser (http://localhost:8080/playground)
{
  searchCourses(params:{ids:[1,2]}){
    books{
      id
      title
    }
  }
}
```

The above request would not return any response and the request would just hang. You can get a response by cancelling the first request and making the same request (without restarting the server). I added [more logs the example project](https://github.com/ExpediaGroup/graphql-kotlin/compare/master...aravind88:add_more_logs?expand=1) and verified that the dataloader is actually not being executed. 

**To Reproduce**

I have a fork additional logs: https://github.com/ExpediaGroup/graphql-kotlin/compare/master...aravind88:add_more_logs?expand=1. 

1. `git clone git@github.com:aravind88/graphql-kotlin.git`
2. `cd graphql-kotlin/examples`
3. `./gradlew server:ktor-server:run`
4. Open a brower to http://localhost:8080/playground and make the following request 
```{
  searchCourses(params:{ids:[1,2]}){
    books{
      id
      title
    }
  }
}
```

You should see the following logs on the first request

```
Jan 03, 2021 4:53:02 PM com.expediagroup.graphql.examples.ktor.schema.models.Course books
SEVERE: start books fetching inside course
Jan 03, 2021 4:53:02 PM com.expediagroup.graphql.examples.ktor.schema.models.Course books
SEVERE: start books fetching inside course
```
Cancelling and re-sending the same request results in the following log:

```
an 03, 2021 4:54:31 PM com.expediagroup.graphql.examples.ktor.schema.models.BookKt$batchBookLoader$1$1 get
SEVERE: starting batch book loader
Jan 03, 2021 4:54:31 PM com.expediagroup.graphql.examples.ktor.schema.models.Course books
SEVERE: start books fetching inside course
Jan 03, 2021 4:54:31 PM com.expediagroup.graphql.examples.ktor.schema.models.Course books
SEVERE: start books fetching inside course
Jan 03, 2021 4:54:31 PM com.expediagroup.graphql.examples.ktor.schema.models.BookKt$batchBookLoader$1$1 get
SEVERE: finished batch book loader
Jan 03, 2021 4:54:31 PM com.expediagroup.graphql.examples.ktor.schema.models.Course books
SEVERE: finished books fetching inside course
Jan 03, 2021 4:54:31 PM com.expediagroup.graphql.examples.ktor.schema.models.Course books
SEVERE: finished books fetching inside course
Jan 03, 2021 4:54:31 PM com.expediagroup.graphql.examples.ktor.schema.models.Course books
SEVERE: finished books fetching inside course
Jan 03, 2021 4:54:31 PM com.expediagroup.graphql.examples.ktor.schema.models.Course books
SEVERE: finished books fetching inside course
```

**Expected behavior**
The first request actually call `dispatch()` on the first request and returns a valid response. 

I am wondering if there is an updated example for Dataloaders I can use or if this is an actual bug.

**Comment by smyrick** [comment]
I haven't gotten a chance to debug the code but running locally I can see that the spring example data loaders are working fine, so I suspect it may have something to do with how the Ktor example data loaders are created and registered

**Comment by aravind88** [comment]
@smyrick I couldn't find an example that uses a `suspend` function along with the dataloader. Let me know if there is a specific example I can take a look at. 

I did some more digging, and I noticed that changing 

`suspend fun books(dataFetchingEnvironment: DataFetchingEnvironment): List<Book>? {`
to 
`fun books(dataFetchingEnvironment: DataFetchingEnvironment): CompletableFuture<List<Book>?> {` 
allowed the dataloader to be dispatched as expected. 

I.e. switching from a `coroutine` based Async Model to a `CompletableFuture` based [Async Model](https://expediagroup.github.io/graphql-kotlin/docs/schema-generator/execution/async-models), fixed the issue.

**Comment by josephlbarnett** [comment]
I ran across this a while ago in my own project, and my understanding is this is the same issue described in upstream java-dataloader here: https://github.com/graphql-java/java-dataloader/pull/46 (or at least very similar to it).  Also found that using `CompletableFuture`s instead of `suspend fun`s for any `FunctionDataFetcher`s that leverage a `DataLoader` to access data was required (but still able to use `suspend fun`s for non-`DataLoader` cases)

The difference here is:
the `CompletableFuture` model:
 * Invoke `DataLoader.load` from the graphql-java datafetcher thread
 * return a Future that gets dispatched when all fetchers done being invoked
vs the `suspend fun` model:
 * Launch a new coroutine and expose as a future 
 * Invoke `DataLoader.load` from that coroutine and return a future to the coroutine, but this never gets dispatched if the fetchers finish before the coroutine finishes execution

basically the extra layer of asynchronicity adds a race between the graphql-java dispatching and the queuing up of the DataLoader future

**Comment by smyrick** [comment]
Reading more into the graphql-java docs, it appears that we can not support this feature and it is a requirement from the graphql-java execution that we return the `CompletableFuture`

https://www.graphql-java.com/documentation/v16/batching/

```java
DataFetcher dataFetcherThatCallsTheDataLoader = new DataFetcher() {
    @Override
    public Object get(DataFetchingEnvironment environment) {
        //
        // Don't DO THIS!
        //
        return CompletableFuture.supplyAsync(() -> {
            String argId = environment.getArgument("id");
            DataLoader<String, Object> characterLoader = environment.getDataLoader("characterLoader");
            return characterLoader.load(argId);
        });
    }
};
```

**Comment by smyrick** [comment]
Reading more into the graphql-java docs, it appears that we can not support this feature and it is a requirement from the graphql-java execution that we return the CompletableFuture

https://www.graphql-java.com/documentation/v16/batching/

```java
DataFetcher dataFetcherThatCallsTheDataLoader = new DataFetcher() {
    @Override
    public Object get(DataFetchingEnvironment environment) {
        //
        // Don't DO THIS!
        //
        return CompletableFuture.supplyAsync(() -> {
            String argId = environment.getArgument("id");
            DataLoader<String, Object> characterLoader = environment.getDataLoader("characterLoader");
            return characterLoader.load(argId);
        });
    }
};
```
